File: packages/domain/src/extraction/evidence.ts
/**
 * Evidence-span registration and exact-quotation verification.
 */
import { assertClaimEvidenceLinkValid, type ClaimEvidenceLink } from '../claims/index.js';
import {
  assertEvidenceRecordValid,
  type EvidenceLocator,
  type EvidenceRecord,
} from '../provenance/index.js';
import type { EvidenceSpan } from './types.js';

function hasExactLocator(locator: EvidenceLocator | undefined): locator is EvidenceLocator {
  if (!locator) return false;
  return Boolean(
    locator.page?.trim() ||
      locator.pages?.trim() ||
      locator.paragraph?.trim() ||
      locator.label?.trim() ||
      locator.uriFragment?.trim() ||
      (Number.isInteger(locator.offsetStart) && Number.isInteger(locator.offsetEnd)),
  );
}

export function assertQuotationAccurate(input: {
  readonly excerpt: string;
  readonly offsetStart: number;
  readonly offsetEnd: number;
  readonly quotation: string;
}): void {
  if (!Number.isInteger(input.offsetStart) || !Number.isInteger(input.offsetEnd)) {
    throw new Error('Evidence span offsets must be integers');
  }
  if (input.offsetStart < 0 || input.offsetEnd <= input.offsetStart) {
    throw new Error('Evidence span offsets must define a non-empty range');
  }
  if (input.offsetEnd > input.excerpt.length) {
    throw new Error('Evidence span exceeds the evidence excerpt');
  }
  if (input.excerpt.slice(input.offsetStart, input.offsetEnd) !== input.quotation) {
    throw new Error('Exact quotation does not match the registered evidence span');
  }
}

export function registerEvidenceSpan(input: {
  readonly id: string;
  readonly evidence: EvidenceRecord;
  readonly offsetStart: number;
  readonly offsetEnd: number;
  readonly exactQuotation: boolean;
  readonly quotation?: string;
}): EvidenceSpan {
  assertEvidenceRecordValid(input.evidence);
  if (!input.id.trim()) throw new Error('Evidence span id is required');
  if (!input.evidence.id.trim()) throw new Error('Evidence id is required');
  if (!input.evidence.excerpt) throw new Error('Evidence span registration requires an excerpt');
  if (!hasExactLocator(input.evidence.locator)) {
    throw new Error('Evidence span registration requires an exact source locator');
  }

  const text = input.evidence.excerpt.slice(input.offsetStart, input.offsetEnd);
  if (!text) throw new Error('Evidence span offsets must define a non-empty range');
  if (input.exactQuotation && input.quotation === undefined) {
    throw new Error('Exact quotations require quotation text');
  }
  assertQuotationAccurate({
    excerpt: input.evidence.excerpt,
    offsetStart: input.offsetStart,
    offsetEnd: input.offsetEnd,
    quotation: input.exactQuotation ? input.quotation! : text,
  });

  const locator = input.evidence.locator;
  return {
    id: input.id,
    evidenceId: input.evidence.id,
    sourceItemId: input.evidence.sourceItemId,
    offsetStart: input.offsetStart,
    offsetEnd: input.offsetEnd,
    text,
    exactQuotation: input.exactQuotation,
    locator: {
      ...(locator.page?.trim() ? { page: locator.page } : {}),
      ...(locator.pages?.trim() ? { pages: locator.pages } : {}),
      ...(locator.paragraph?.trim() ? { paragraph: locator.paragraph } : {}),
      ...(locator.offsetStart !== undefined && Number.isInteger(locator.offsetStart)
        ? { offsetStart: locator.offsetStart }
        : {}),
      ...(locator.offsetEnd !== undefined && Number.isInteger(locator.offsetEnd)
        ? { offsetEnd: locator.offsetEnd }
        : {}),
      ...(locator.label?.trim() ? { label: locator.label } : {}),
      ...(locator.uriFragment?.trim() ? { uriFragment: locator.uriFragment } : {}),
    },
  };
}

export function evidenceLinkQualifies(
  link: ClaimEvidenceLink,
  spans: readonly EvidenceSpan[],
  expectedValue?: string,
): boolean {
  assertClaimEvidenceLinkValid(link);
  return (
    link.role === 'supporting' &&
    link.credible &&
    link.directness > 0 &&
    link.extractionQuality > 0 &&
    (expectedValue === undefined || link.assertedValue?.trim() === expectedValue.trim()) &&
    spans.some((span) => span.evidenceId === link.evidenceId)
  );
}

File: workers/research/src/black_book_research/extraction/evidence.py
"""Exact-quotation checks and immutable evidence-span registration."""

from __future__ import annotations

from collections.abc import Mapping

from .models import EvidenceSpan


def assert_quotation_accurate(
    excerpt: str, offset_start: int, offset_end: int, quotation: str
) -> None:
    if (
        isinstance(offset_start, bool)
        or isinstance(offset_end, bool)
        or not isinstance(offset_start, int)
        or not isinstance(offset_end, int)
    ):
        raise ValueError("Evidence span offsets must be integers")
    if offset_start < 0 or offset_end <= offset_start:
        raise ValueError("Evidence span offsets must define a non-empty range")
    if offset_end > len(excerpt):
        raise ValueError("Evidence span exceeds the evidence excerpt")
    if excerpt[offset_start:offset_end] != quotation:
        raise ValueError("Exact quotation does not match the registered evidence span")


def register_evidence_span(
    *,
    span_id: str,
    evidence_id: str,
    source_item_id: str,
    excerpt: str,
    locator: Mapping[str, str],
    offset_start: int,
    offset_end: int,
    exact_quotation: bool,
    quotation: str | None = None,
) -> EvidenceSpan:
    if not span_id.strip() or not evidence_id.strip() or not source_item_id.strip():
        raise ValueError("Evidence span, evidence, and source item ids are required")
    exact_locator = tuple(
        sorted((key, value) for key, value in locator.items() if value.strip())
    )
    if not exact_locator:
        raise ValueError("Evidence span registration requires an exact source locator")
    text = excerpt[offset_start:offset_end]
    if exact_quotation and quotation is None:
        raise ValueError("Exact quotations require quotation text")
    assert_quotation_accurate(
        excerpt,
        offset_start,
        offset_end,
        quotation if exact_quotation and quotation is not None else text,
    )
    return EvidenceSpan(
        id=span_id,
        evidence_id=evidence_id,
        source_item_id=source_item_id,
        offset_start=offset_start,
        offset_end=offset_end,
        text=text,
        exact_quotation=exact_quotation,
        locator=exact_locator,
    )
